### DISCUSSION TITLE
Problem with delay command within rule

### DISCUSSION THREAD

**Discussion by ggggh** [arendst/Tasmota#11745] [Q&A] [answered]
I'm having a weird problem with a rule, and wondering if it's a bug or more likely if I've missed something in the docs:

```
Rule2 
ON Switch4#state=1 do Backlog publish /home/state/water_pump ON;POWER3 ON;Delay 150;POWER3 OFF ENDON
ON Switch4#state=0 do publish /home/state/water_pump OFF ENDON
```

First condition of the rule (state=1) triggers when water reaches a certain level within a water tank.  It notifies me by publishing an mqtt message, and then should switch on the pump (Power3) for 15 seconds, to drain the water out and lower the water level.
Around one second after switching the pump on (Power3 ON), the second rule condition triggers (state=0), as the water level drops quite quickly.  This is also expected, and notifies me the level is back under the limit, but it has the unexpected effect of apparently cancelling the 15 second delay on the backlog of the prior state=1 rule execution (which is still running), and running the POWER3 OFF immediately.  Note it doesn't cancel the backlog, it just doesn't wait for the 15 seconds - basically interrupts the running delay.

Is this expected and documented behaviour?  I've not found anything mentioned about this.

If I run that full rule mentioned above, I get this output below (note all within a second, regardless of the 15 sec delay):

```
23:32:26 RUL: SWITCH4#STATE=1 performs "Backlog publish /home/state/water_pump ON;POWER3 ON;Delay 150;POWER3 OFF;"
23:32:26 MQT: /home/state/water_pump = ON
23:32:26 MQT: /home/switches/sonoff4ch_2/RESULT = {"POWER3":"ON"}
23:32:26 MQT: /home/switches/sonoff4ch_2/POWER3 = ON
23:32:26 MQT: /home/switches/sonoff4ch_2/RESULT = {"Delay":150}
23:32:27 RUL: SWITCH4#STATE=0 performs "Publish /home/state/water_pump OFF;"
23:32:27 MQT: /home/state/water_pump = OFF;
23:32:27 MQT: /home/switches/sonoff4ch_2/RESULT = {"POWER3":"OFF"}
23:32:27 MQT: /home/switches/sonoff4ch_2/POWER3 = OFF
```

If instead I remove the second condition (the "state=0" line) and only include the "state=1" line in the rule, it works fine and takes the expected 15 seconds:

```
18:46:18 RUL: SWITCH4#STATE=1 performs "Backlog publish /home/state/water_pump ON;POWER3 ON;Delay 150;POWER3 OFF;"
18:46:18 MQT: /home/state/water_pump = ON
18:46:18 MQT: /home/switches/sonoff4ch_2/RESULT = {"POWER3":"ON"}
18:46:18 MQT: /home/switches/sonoff4ch_2/POWER3 = ON
18:46:18 MQT: /home/switches/sonoff4ch_2/RESULT = {"Delay":150}
18:46:18 MQT: /home/switches/sonoff4ch_2/RESULT = {"POWER4":"OFF"}
18:46:18 MQT: /home/switches/sonoff4ch_2/POWER4 = OFF
18:46:33 MQT: /home/switches/sonoff4ch_2/RESULT = {"POWER3":"OFF"}
18:46:33 MQT: /home/switches/sonoff4ch_2/POWER3 = OFF
```

It seems to me it is an issue with two instances of a rule running in parallel; in general they work, but it seems the second instance incorrectly cancels a "delay" running within the first instance.

**Comment by sfromis** [comment, marked as answer]
If you want reliable timing in rules, do not use Delay. RuleTimer recommended. It is correct that other commands executed can cut short a pending Delay.

**Reply by meingraham** [reply]
Also, since your use case appears to be to turn off the relay after a set time, you could set up `PulseTime` and the relay will turn off automatically after that set period without you or a rule having to issue the `OFF` command.
